llvm/lib/Target/AIE/ReservedRegsLICM.cpp

```diff
@@ -0,0 +1,381 @@
+}
+
+BitVector ReservedRegsLICM::collectLoopReservedLiveins(const MachineLoop &L) {
+  const MachineBasicBlock &Header = *L.getHeader();
+  LivePhysRegs LiveRegs(*TRI);
+
+  // Conservativaly assume reserved reserved regs are all liveouts
+  for (MCPhysReg PhysReg : MRI->getReservedRegs().set_bits()) {
+    if (MRI->canSimplifyPhysReg(PhysReg)) {
+      LiveRegs.addReg(PhysReg);
+    }
+  }
+
+  // Traverse instructions to remove defs
+  for (const MachineInstr &MI : reverse(Header))
+    LiveRegs.stepBackward(MI);
+
+  BitVector ReservedLiveins(TRI->getNumRegs());
+  for (MCRegister Reg : LiveRegs) {
+    if (MRI->isReserved(Reg)) {
+      ReservedLiveins.set(Reg);
+    }
+  }
+  return ReservedLiveins;
+}
+
+/// Walk the specified region of the CFG and hoist loop invariants out to the
+/// preheader.
+void ReservedRegsLICM::runOnLoop(MachineLoop &L) {
+  MachineBasicBlock *Preheader = L.getLoopPreheader();
+  MachineBasicBlock *ExitBlock = L.getExitBlock();
+  if (!Preheader || !ExitBlock) {
+    LLVM_DEBUG(dbgs() << "  Loop has no preheader or single exit block.\n");
+    return;
+  }
+
+  // TODO: Handle simple multi-BB loops.
+  if (L.getNumBlocks() != 1) {
+    LLVM_DEBUG(dbgs() << "  Loop has multiple blocks.\n");
+    return;
+  }
+
+  BitVector ReservedLiveins = collectLoopReservedLiveins(L);
+  processForExitSink(L, ReservedLiveins);
+  processForPreheaderHoist(L, ReservedLiveins);
+}
+
+void ReservedRegsLICM::processForExitSink(MachineLoop &L,
+                                          const BitVector &ReservedLiveins) {
+  RegDefMap PhysRegChanged(*TRI);
+  LivePhysRegs LiveRegs(*TRI);
+  Candidates.clear();
+
+  // Walk the entire region, track defs for each register, and
+  // collect potential LICM candidates.
+  assert(L.getNumBlocks() == 1 && L.getLoopLatch());
+  for (MachineInstr &MI : reverse(*L.getLoopLatch())) {
+    CandidateInfo *CandInfo = getCandidateInfo(MI);
+
+    // First time we meet a reserved reg definition while iterating upwards.
+    // If that def is not a loop livein and it isn't used in this block either,
+    // then one can move the instruction to the exit BB of the loop.
+    if (CandInfo && !PhysRegChanged.hasChanged(CandInfo->DefinedReg) &&
+        !ReservedLiveins.test(CandInfo->DefinedReg) &&
+        !LiveRegs.contains(CandInfo->DefinedReg)) {
+      assert(!CandInfo->HoistCandidate);
+      CandInfo->HoistCandidate = &MI;
+    }
+
+    // Track physregs that were changed/used by MI, and what's now live.
+    PhysRegChanged.addChangedRegs(MI);
+    LiveRegs.stepBackward(MI);
+  }
+
+  for (auto &[Reg, CandInfo] : Candidates) {
+    Changed |= trySinkToExitBlock(CandInfo, L);
+  }
+}
+
+bool ReservedRegsLICM::trySinkToExitBlock(const CandidateInfo &Cand,
+                                          MachineLoop &L) {
+  if (isLoopInvariantInst(Cand, L)) {
+    MachineBasicBlock *InsertMBB = L.getExitBlock();
+    assert(InsertMBB);
+    hoist(Cand, InsertMBB->getFirstNonPHI(), *InsertMBB);
```

Comment:
As we are in `trySinkToExitBlock` it is a good idea to rename `hoist` to a different name, like `hoistOrSink` os just `moveInstruction`. You ca also pass directly the instruction instead of Cand.